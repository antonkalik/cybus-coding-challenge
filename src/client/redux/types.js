export const UPDATE_STORE = 'UPDATE_STORE';
export const RESET_STORE = 'RESET_STORE';
