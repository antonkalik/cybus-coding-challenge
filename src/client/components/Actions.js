import React from 'react';
import { filterObj } from '../utilities';
import start from '../res/actions/start.svg';
import stop from '../res/actions/stop.svg';
import restart from '../res/actions/restart.svg';
import remove from '../res/actions/remove.svg';

export default function Actions({ item, index, updateStore }) {
  const actions = { remove, restart, stop, start };
  const getActionKeys = (actions, status) => {
    if (status === 'dead') {
      return ['remove'];
    }
    const keysForRemove = status === 'up' ? 'start' : 'stop';
    const getObj = filterObj(actions, keysForRemove);
    return Object.keys(getObj);
  };
  return (
    <div className="actions">
      {getActionKeys(actions, item.status).map(key => (
        <div className="action" key={key}>
          <div className="tooltip">{key}</div>
          <img
            onClick={() => {
              updateStore({ shareData: { key, index, id: item.id }, modal: true });
            }}
            src={actions[key]}
            alt={key}
          />
        </div>
      ))}
    </div>
  );
}
