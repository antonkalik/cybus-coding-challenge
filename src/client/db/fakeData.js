export default {
  data: {
    images: [],
    containers: [],
  },
};
