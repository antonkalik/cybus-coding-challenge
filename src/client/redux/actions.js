import { UPDATE_STORE, RESET_STORE, UPDATE_CONTAINER, UPDATE_TAB, RESET_SEARCH } from './types';

export const actionUpdateStore = (payload = {}) => {
  return {
    type: UPDATE_STORE,
    payload,
  };
};

export const actionResetStore = () => {
  return {
    type: RESET_STORE,
  };
};

export const actionUpdateContainer = (container, index) => {
  return {
    type: UPDATE_CONTAINER,
    container,
    index,
  };
};

export const actionUpdateTab = currentTab => {
  return {
    type: UPDATE_TAB,
    currentTab,
  };
};
