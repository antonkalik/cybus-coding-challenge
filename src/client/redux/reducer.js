import { createReducer } from '../utilities';
import { RESET_STORE, UPDATE_STORE } from './types';

const initialState = {
  isLoggedIn: false,
  userName: '',
  currentTab: 'images',
  images: [],
  containers: [],
};

export default createReducer(initialState, {
  [UPDATE_STORE]: (state, { payload }) => {
    return {
      ...state,
      ...payload,
    };
  },
  [RESET_STORE]: () => {
    return initialState;
  },
});
