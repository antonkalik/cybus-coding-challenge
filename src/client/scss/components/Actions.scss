.actions {
  display: flex;
  justify-content: flex-end;
  align-items: center;
  > div.action {
    position: relative;
    display: flex;
    justify-content: flex-end;
    height: 30px;
    width: 30px;
    margin: 4px;
    .tooltip {
      position: absolute;
      display: none;
      z-index: 99;
      border: 2px solid $black;
      background: $light;
      top: -50px;
      left: -20px;
      padding: 10px;
      border-radius: 4px;
    }
    &:hover {
      .tooltip {
        display: block;
      }
    }
    img {
      position: absolute;
      padding: 2px;
      border: 2px solid $black;
      box-shadow: 2px 2px $black;
      border-radius: 8px;
      background: white;
      cursor: pointer;
      width: auto;
      height: 24px;
      top: -2px;
      right: 0;

      &:hover {
        top: 0;
        right: -2px;
        box-shadow: none;
      }
    }
  }
}
