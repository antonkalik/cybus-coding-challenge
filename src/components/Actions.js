import React from 'react';
import { getActionKeys } from '../utilities';

export default function Actions({ item, index, updateStore }) {
  const path = 'res/actions/';
  const actions = ['remove', 'restart', 'stop', 'start'].reduce((a, b) => {
    a[b] = `${path + b}.svg`;
    return a;
  }, {});

  return (
    <div className="actions">
      {getActionKeys(actions, item.status).map(key => (
        <div className="action" key={key}>
          <div className="tooltip">{key}</div>
          <img
            onClick={() => {
              updateStore({ shareData: { key, index, id: item.id }, modal: true });
            }}
            src={actions[key]}
            alt={key}
          />
        </div>
      ))}
    </div>
  );
}
