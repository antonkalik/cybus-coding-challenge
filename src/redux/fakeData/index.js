export const fakeData = {
  images: [
    {
      repository: 'johnwayne/client-side-server',
      id: '7sdkf842j7skcc',
      tag: 'latest',
      created: '12 hours ago',
      size: '1.2GB',
    },
    {
      repository: 'johnwayne/my-nginx',
      id: 'tuw6394kkd3mvd',
      tag: 'latest',
      created: '9 days ago',
      size: '725MB',
    },
    {
      repository: 'node',
      id: 'eu33h2djf0ok2mv',
      tag: '10',
      created: '24 days ago',
      size: '903MB',
    },
    {
      repository: 'ssr-sql',
      id: '6hdh366dhc2d2s',
      tag: 'latest',
      created: '5 days ago',
      size: '1.5GB',
    },
  ],
  containers: [
    {
      id: '7ycn4sk3kv8s',
      image: 'johnwayne/my-nginx',
      created: '24 days ago',
      status: 'up',
      names: 'modest_mayer',
    },
    {
      id: '2ydkcm84jdjdv',
      image: 'johnwayne/client-side-server',
      created: '9 days ago',
      status: 'up',
      names: 'compassionate_newton',
    },
    {
      id: '4bb99skvisdsiis',
      image: 'johnwayne/client-side-server',
      created: '5 seconds ago',
      status: 'paused',
      names: 'optimistic_mendel',
    },
    {
      id: 'idj73jhv7dkjlla',
      image: 'node',
      created: '31 hours ago',
      status: 'up',
      names: 'romantic_boyd',
    },
    {
      id: '2wi3radsgc23sfnb',
      image: 'johnwayne/client-side-server',
      created: '29 minutes ago',
      status: 'paused',
      names: 'nice_liskov',
    },
    {
      id: '8vjsj22aawv3or',
      image: 'johnwayne/my-node',
      created: '18 hours ago',
      status: 'exited',
      names: 'boring_wozniak',
    },
    {
      id: 'svjjsmkm2j4mss',
      image: 'ssr-sql',
      created: '15 days ago',
      status: 'dead',
      names: 'agitated_darwin',
    },
  ],
};

export const headers = {
  images: Object.keys(fakeData.images[0]),
  containers: Object.keys(fakeData.containers[0]),
};
