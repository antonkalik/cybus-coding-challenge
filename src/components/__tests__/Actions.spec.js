import React from 'react';
import { shallow } from 'enzyme';
import Actions from '../Actions';
import { fakeData } from '../../redux/fakeData';

const item = fakeData.containers[0];
const itemWithDeadStatus = fakeData.containers.find(it => it.status === 'dead');

describe('<Actions /> test', () => {
  test('should show 3 actions', () => {
    const component = shallow(<Actions item={item} index={0} updateStore={() => {}} />);
    expect(component.hasClass('actions')).toBeTruthy();
    expect(component.find('.action')).toHaveLength(3);
    expect(component).toMatchSnapshot();
  });
  test('should show only remove action', () => {
    const component = shallow(
      <Actions item={itemWithDeadStatus} index={0} updateStore={() => {}} />
    );
    expect(component.hasClass('actions')).toBeTruthy();
    expect(component.find('.action')).toHaveLength(1);
    expect(component).toMatchSnapshot();
  });
});
