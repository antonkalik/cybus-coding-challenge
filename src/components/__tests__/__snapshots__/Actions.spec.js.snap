// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Actions /> test should show 3 actions 1`] = `ShallowWrapper {}`;

exports[`<Actions /> test should show only remove action 1`] = `ShallowWrapper {}`;
